src/lib/scheduler/topology.cpp

```diff
@@ -71,23 +71,32 @@ void Topology::_init_numa_topology(uint32_t max_num_cores) {
 
   auto max_node = numa_max_node();
   auto num_configured_cpus = static_cast<CpuID>(numa_num_configured_cpus());
-  auto cpu_bitmask = numa_allocate_cpumask();
+
+  // We take the CPU affinity (set, e.g., by numactl) of our process into account.
+  // Otherwise, we would always start with the first CPU, even if a specific NUMA node was selected.
+  auto affinity_cpu_bitmask = numa_allocate_cpumask();
+  numa_sched_getaffinity(0, affinity_cpu_bitmask);
+
+  auto this_node_cpu_bitmask = numa_allocate_cpumask();
   auto core_count = uint32_t{0};
 
   for (auto node_id = 0; node_id <= max_node; node_id++) {
     if (max_num_cores == 0 || core_count < max_num_cores) {
       auto cpus = std::vector<TopologyCpu>();
 
-      numa_node_to_cpus(node_id, cpu_bitmask);
+      numa_node_to_cpus(node_id, this_node_cpu_bitmask);
 
       for (CpuID cpu_id{0}; cpu_id < num_configured_cpus; ++cpu_id) {
-        if (numa_bitmask_isbitset(cpu_bitmask, cpu_id)) {
+        const auto cpu_is_part_of_node = numa_bitmask_isbitset(this_node_cpu_bitmask, cpu_id);
+        const auto cpu_is_part_of_affinity = numa_bitmask_isbitset(affinity_cpu_bitmask, cpu_id);
+        if (cpu_is_part_of_node && cpu_is_part_of_affinity) {
           if (max_num_cores == 0 || core_count < max_num_cores) {
             cpus.emplace_back(TopologyCpu(cpu_id));
             _num_cpus++;
```

Comment:
`_num_cpus` does actually describe cores, right?

Reply:
Ah, I just see that libnuma uses `cpu` for what I would call a core.
It's somewhat confusing that we talk about CPUs and cores and then increase `_num_cpus` if the core/cpu is used while `core_count` is always increased.

The check in line 84 ensure that we don't use too many cores, right? But `core_count` is increased no matter whether the cpu is actually used due to the affinity map.

Reply:
This is hard to do right. We like to talk about cores, libnuma calls them cpus, so does `lscpu`. The topology as at the boundary of the two. However, I don't expect this to really cause an issue in the future.

> The check in line 84 ensure that we don't use too many cores, right?

Correct

> But core_count is increased no matter whether the cpu is actually used due to the affinity map.

Incorrect.